File: typescript/private_key/5_limit_buy_order.ts
/**
 * 5_limit_order.ts — GTC limit BUY 2 ticks below best bid → confirm → cancel.
 *
 * EOA-based: signatureType=0, pUSD held by EOA (not Safe).
 *
 * Run:
 *   npm run 5:limit
 */

import "dotenv/config";
import { createWalletClient, createPublicClient, http, erc20Abi, formatUnits } from "viem";
import { polygon } from "viem/chains";
import { privateKeyToAccount } from "viem/accounts";
import { ClobClient, Side, OrderType } from "@polymarket/clob-client-v2";

const PRIVATE_KEY = process.env.PRIVATE_KEY!;
const CLOB_HOST   = process.env.CLOB_V2_BASE_URL ?? "https://clob-v2.polymarket.com";
const RPC_URL     = process.env.POLYGON_RPC_URL  ?? "https://polygon-bor-rpc.publicnode.com";
const EVENT_ID    = "73106";
const PUSD        = "0xC011a7E12a19f7B1f670d46F03B03f3342E82DFB" as `0x${string}`;

// ── Clients ───────────────────────────────────────────────────────────────────

const account      = privateKeyToAccount(`0x${PRIVATE_KEY}` as `0x${string}`);
const walletClient = createWalletClient({ account, chain: polygon, transport: http(RPC_URL) });
const publicClient = createPublicClient({ chain: polygon, transport: http(RPC_URL) });
const EOA          = account.address;

console.log("EOA  :", EOA);
console.log("CLOB :", CLOB_HOST, "\n");

// ── Helpers ───────────────────────────────────────────────────────────────────

async function printBook(label: string, tokenId: string, highlightPrice?: number) {
  const book = await (client as any).getOrderBook(tokenId);
  const bids = [...(book?.bids ?? [])].sort((a: any, b: any) => parseFloat(b.price) - parseFloat(a.price));
  const asks = [...(book?.asks ?? [])].sort((a: any, b: any) => parseFloat(a.price) - parseFloat(b.price));
  const mid  = bids[0] && asks[0] ? (parseFloat(bids[0].price) + parseFloat(asks[0].price)) / 2 : null;
  const mark = (p: string) => highlightPrice !== undefined && parseFloat(p) === highlightPrice ? " ◄ our order" : "";

  console.log(`\n──── ${label} ${"─".repeat(Math.max(0, 38 - label.length))}`);
  [...asks.slice(0, 5)].reverse().forEach((a: any, i, arr) => {
    console.log(`  ask ${arr.length - i}  $${parseFloat(a.price).toFixed(4)}  sz: ${a.size}${mark(a.price)}`);
  });
  console.log(mid ? `  ── mid ──  $${mid.toFixed(4)}` : "  ── mid ──  (one side empty)");
  bids.slice(0, 5).forEach((b: any, i: number) => {
    console.log(`  bid ${i + 1}  $${parseFloat(b.price).toFixed(4)}  sz: ${b.size}${mark(b.price)}`);
  });
  console.log(`${"─".repeat(44)}  asks: ${asks.length}  bids: ${bids.length}`);
  return { bids, asks };
}

// ── Step 1: Init ClobClient V2 (EOA) ─────────────────────────────────────────

const tempClient = new ClobClient({ host: CLOB_HOST, chain: 137, signer: walletClient as never, signatureType: 0 });
const creds      = await (tempClient as any).createOrDeriveApiKey();
console.log("API key :", creds.key, "✓");

const client = new ClobClient({
  host: CLOB_HOST, chain: 137, signer: walletClient as never, signatureType: 0,
  creds: { key: creds.key, secret: creds.secret, passphrase: creds.passphrase },
});
console.log("ClobClient V2 initialized ✓");

// ── Step 2: Market info ───────────────────────────────────────────────────────

const event   = await fetch(`https://gamma-api.polymarket.com/events/${EVENT_ID}`).then(r => r.json()) as any;
const condId  = event.markets?.[0]?.conditionId as string;
const info    = await (client as any).getClobMarketInfo(condId);
const tokens  = (info?.t ?? []) as { t: string; o: string }[];
const tokenId = tokens.find(t => t.o === "Yes")?.t ?? "";
const tickSize = (info?.mts?.toString() ?? "0.01") as "0.1" | "0.01" | "0.001" | "0.0001";
const negRisk = info?.nr ?? false;

console.log(`Market: tick=${tickSize}  negRisk=${negRisk}  token=${tokenId.slice(0, 16)}…`);

// ── Step 3: Book BEFORE ───────────────────────────────────────────────────────

const { bids } = await printBook("BEFORE order", tokenId);

if (bids.length < 3) { console.error("Not enough bid levels"); process.exit(1); }

const price = parseFloat((bids[2] as any).price);
console.log(`\nTarget price (bid-3): $${price}`);

// ── Step 4: pUSD balance ──────────────────────────────────────────────────────

const rawBal  = await publicClient.readContract({ address: PUSD, abi: erc20Abi, functionName: "balanceOf", args: [EOA] });
const balance = parseFloat(formatUnits(rawBal, 6));
console.log(`pUSD balance: $${balance.toFixed(6)}`);

if (balance < 1) { console.error("⚠ Not enough pUSD — run 3_wrap.ts first."); process.exit(1); }

const BUFFER = 1;
const size   = Math.floor(((balance - BUFFER) / price) * 100) / 100;
console.log(`Order size : ${size} shares  ($${(balance - BUFFER).toFixed(2)} ÷ $${price})`);

// ── Step 5: Place GTC limit BUY ───────────────────────────────────────────────

console.log(`\nPlacing GTC BUY: ${size} shares @ $${price}...`);
const resp = await client.createAndPostOrder(
  { tokenID: tokenId, price, size, side: Side.BUY },
  { tickSize, negRisk },
  OrderType.GTC,
);
console.log(`  Order ID : ${resp.orderID}`);
console.log(`  Status   : ${resp.status}`);
if (!resp.orderID) { console.error("Order rejected:", (resp as any).errorMsg); process.exit(1); }

// ── Step 6: Book WHILE live ───────────────────────────────────────────────────

await printBook("WHILE order is live", tokenId, price);

// ── Step 7: Cancel ────────────────────────────────────────────────────────────

console.log(`\nCancelling ${resp.orderID!.slice(0, 18)}…`);
const cancel = await client.cancelOrder({ orderID: resp.orderID! });
console.log(cancel?.canceled?.includes(resp.orderID!) ? "✓ Cancelled" : `Cancel resp: ${JSON.stringify(cancel)}`);

// ── Step 8: Book AFTER cancel ─────────────────────────────────────────────────

await printBook("AFTER cancel", tokenId);

File: python/private_key/5_limit_buy_order.py
"""
5_limit_buy_order.py — GTC BUY 2 ticks below best bid → confirm in book → cancel.

Run:
    python 5_limit_buy_order.py

Requires: PRIVATE_KEY, CLOB_V2_BASE_URL, POLYGON_RPC_URL in .env
pUSD held by EOA (not a Safe). Min order size = 5 shares.
"""

import os
import requests
from dotenv import load_dotenv
from web3 import Web3
from py_clob_client_v2.client import ClobClient
from py_clob_client_v2.clob_types import OrderArgsV2, PartialCreateOrderOptions, OrderType, BalanceAllowanceParams, AssetType, OrderPayload

load_dotenv(dotenv_path=os.path.join(os.path.dirname(__file__), ".env"))

PRIVATE_KEY = os.environ["PRIVATE_KEY"]
CLOB_HOST   = os.environ.get("CLOB_V2_BASE_URL", "https://clob-v2.polymarket.com")
RPC_URL     = os.environ.get("POLYGON_RPC_URL", "https://polygon-bor-rpc.publicnode.com")
EVENT_ID    = "73106"
PUSD        = Web3.to_checksum_address("0xC011a7E12a19f7B1f670d46F03B03f3342E82DFB")

ERC20_ABI = [{"name": "balanceOf", "type": "function", "stateMutability": "view",
              "inputs": [{"name": "account", "type": "address"}],
              "outputs": [{"name": "", "type": "uint256"}]}]

w3  = Web3(Web3.HTTPProvider(RPC_URL))
eoa = w3.eth.account.from_key("0x" + PRIVATE_KEY).address

# ── Init client ───────────────────────────────────────────────────────────────
temp_client = ClobClient(CLOB_HOST, chain_id=137, key=PRIVATE_KEY, signature_type=0)
creds       = temp_client.create_or_derive_api_key()
client      = ClobClient(CLOB_HOST, chain_id=137, key=PRIVATE_KEY, creds=creds, signature_type=0)
print(f"EOA     : {eoa}")
print(f"API key : {creds.api_key} ✓\n")

# ── Sync on-chain balances/allowances to CLOB server ─────────────────────────
client.update_balance_allowance(BalanceAllowanceParams(asset_type=AssetType.COLLATERAL))
print("Balance/allowance synced ✓\n")

# ── Market info ───────────────────────────────────────────────────────────────
event     = requests.get(f"https://gamma-api.polymarket.com/events/{EVENT_ID}").json()
cond_id   = event["markets"][0]["conditionId"]
info      = client.get_clob_market_info(cond_id)
tokens    = info.get("t", [])
token_id  = next((t["t"] for t in tokens if t["o"] == "Yes"), "")
tick_size = str(info.get("mts", "0.01"))
neg_risk  = info.get("nr", False)
print(f"Market: tick={tick_size}  negRisk={neg_risk}  token={token_id[:16]}…")

# ── Order book — 2 ticks below best bid ───────────────────────────────────────
book = client.get_order_book(token_id)
bids = sorted(book["bids"], key=lambda b: float(b["price"]), reverse=True)
if len(bids) < 3:
    print("Not enough bid levels"); exit(1)

price = float(bids[2]["price"])

# ── pUSD balance ──────────────────────────────────────────────────────────────
raw_bal = w3.eth.contract(address=PUSD, abi=ERC20_ABI).functions.balanceOf(eoa).call()
balance = raw_bal / 1e6
size    = round((balance - 1) / price, 2)
size    = max(size, 0)

print(f"Balance : ${balance:.4f} pUSD")
print(f"Placing GTC BUY: {size} shares @ ${price}")

if size <= 0:
    print("⚠ Insufficient pUSD balance"); exit(1)

# ── Place GTC limit BUY ───────────────────────────────────────────────────────
resp = client.create_and_post_order(
    OrderArgsV2(token_id=token_id, price=price, size=size, side="BUY"),
    PartialCreateOrderOptions(tick_size=tick_size, neg_risk=neg_risk),
    OrderType.GTC,
)
order_id = resp.get("orderID") if isinstance(resp, dict) else resp.orderID
status   = resp.get("status")  if isinstance(resp, dict) else resp.status
print(f"\n✓ Order placed : {order_id}  status: {status}")

# ── Cancel (cleanup) ──────────────────────────────────────────────────────────
cancel = client.cancel_order(OrderPayload(orderID=order_id))
canceled = cancel.get("canceled", []) if isinstance(cancel, dict) else []
if order_id in canceled:
    print(f"✓ Cancelled    : {order_id}")
else:
    print(f"Cancel response: {cancel}")
